### ISSUE TITLE
Propagate TLS certs updates to admin / frontend / matching / history clients

### ISSUE THREAD

**Issue by wxing1292** [temporalio/temporal#4918]
## Expected Behavior
When TLS certs are updated, properly propagate the cert changes to admin / frontend / matching / history clients

## Actual Behavior
Seems that 
[GetRemoteAdminClient](https://github.com/temporalio/temporal/blob/main/client/client_bean.go#L177)
[GetRemoteFrontendClient](https://github.com/temporalio/temporal/blob/main/client/client_bean.go#L226)
both functions will cache created admin / frontend client forever, even after TLS certs are updated

**Comment by dnr** [comment]
Have you verified this? I just walked through the code and it looks like it's doing all the right things to refresh certs everywhere, including remote admin/frontend connections. (Just because the grpc client isn't changing doesn't mean the underlying tls connection isn't reloading certs.)

**Comment by wxing1292** [comment]
I have verified that TLS cert refresh does not work; Cert is updated in k8s but client cannot talk to remote; rolling restart fixed the issue; this is probably the second time we have seems this issue;

from my reading the clients are created and cached in memory; never refreshed (see above link) 
So can you point me to the logic which updates the clients?

**Comment by wxing1292** [comment]
> Just to be sure, did you set `refreshInterval` within `global.tls`? It defaults to 0 (no refresh)
> 
> Reloading certs happens way below creation of grpc clients. See e.g. https://github.com/temporalio/temporal/blob/main/common/auth/tls_config_helper.go#L65-L67 https://github.com/temporalio/temporal/blob/main/common/rpc/encryption/local_store_tls_provider.go#L318 It sets callbacks and the Go TLS stack calls them when appropriate, you don't need to recreate the grpc client. The refresh is here: https://github.com/temporalio/temporal/blob/main/common/rpc/encryption/local_store_cert_provider.go#L81-L85 https://github.com/temporalio/temporal/blob/main/common/rpc/encryption/local_store_cert_provider.go#L523

cc @dynajoe

**Comment by dnr** [comment]
I tried to reproduce this in a bunch of ways:

1. I set up two development servers (A and B) connected to each other with one CA and each one using a different cert.
`refreshInterval` was set to 10s on both.
After setting up the connections, I replaced B's certs with a cert signed by a different CA and then restarted the servers.
As expected, B started logging errors connecting to A (which was rejecting the cert signed by the wrong CA).
(I'm not entirely sure why A didn't log errors connection to B here.)
Without restarting B, I replaced its cert with a good one again.
Eventually (after both worker and history reloaded certs), it stopped logging errors.
I confirmed (by adding printfs) that it does not do another call to `grpc.Dial` after startup, the cert reload was enough.

2. Then I tried replacing B's good cert with a bad one again, without restarting the server.
Things continued working for several minutes using the established connection, but within 5 minutes (maybe a total connection length limit?), it started logging errors.
I replaced B's cert with a good one (again without restarting), and both eventually stopped logging errors.

4. Last, I replaced B's cert with one that expires in 5 minutes and started both clusters.
After 5 minutes, I didn't see any errors immediately, but within 5 more minutes it started logging errors (client: `connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2023-10-16T13:21:01-07:00 is after 2023-10-16T20:16:47Z"`, server: `connection error: desc = "error reading server preface: remote error: tls: expired certificate"`).
I replaced B's cert with a good one again, and both eventually stopped logging errors.

So it looks like cert reloading does seem to work in a local test setup.
